In [1]:
# Import modules 
import pandas as pd
import numpy as np
import datetime
import time

Read in the data web scraped from spotrac.com
The data frame consists of contract information of MLB free agents ranging from 2012 to 2020. 

In [2]:
FA_list = pd.read_csv("FA_list.csv")

In [3]:
Names = list(FA_list["Name"])

In [4]:
Names[0]

'Albert Pujols'

In [5]:
FA_list.head()

Name Position  Age From   To  Years       Dollars Average_Salary  \
0   Albert Pujols       DH   32  STL  LAA     10  $240,000,000    $24,000,000   
1  Prince Fielder       DH   28  MIL  DET      9  $214,000,000    $23,777,778   
2      Jose Reyes       SS   29  NYM  MIA      6  $106,000,000    $17,666,667   
3     C.J. Wilson       SP   31  TEX  LAA      5   $77,500,000    $15,500,000   
4    Mark Buehrle       SP   33  CHW  MIA      4   $58,000,000    $14,500,000   

   FA_year  
0     2012  
1     2012  
2     2012  
3     2012  
4     2012

In [6]:
FA_list.shape

(1118, 9)

FA_list has 1,118 rows

## Join bbrefID from baseballdatabank
Adding baseball reference ID from basebll databank to the free agents so we could further web scrape player statistics from baseballreference.com   <br/>
__[Baseball databank](http://www.seanlahman.com/baseball-archive/statistics/)__ is a compilation of historical baseball data in a convenient, tidy format, distributed under Open Data terms.

bbrefID takes the first two letters of a players first name + the first five letters of a player's last name + two number to form. <br/>
For example: bbrefID for Albert Pujos would be pujolal01.

In [7]:
# read in data from baseball databank
People = pd.read_csv("C:\\Users\\iamia\\Desktop\\MLB Free Agent\\baseballdatabank-master\\core\\People.csv")

In [8]:
People.head(10)

playerID  birthYear  birthMonth  birthDay birthCountry birthState  \
0  aardsda01     1981.0        12.0      27.0          USA         CO   
1  aaronha01     1934.0         2.0       5.0          USA         AL   
2  aaronto01     1939.0         8.0       5.0          USA         AL   
3   aasedo01     1954.0         9.0       8.0          USA         CA   
4   abadan01     1972.0         8.0      25.0          USA         FL   
5   abadfe01     1985.0        12.0      17.0         D.R.  La Romana   
6  abadijo01     1850.0        11.0       4.0          USA         PA   
7  abbated01     1877.0         4.0      15.0          USA         PA   
8  abbeybe01     1869.0        11.0      11.0          USA         VT   
9  abbeych01     1866.0        10.0      14.0          USA         NE   

      birthCity  deathYear  deathMonth  deathDay  ...     nameLast  \
0        Denver        NaN         NaN       NaN  ...      Aardsma   
1        Mobile        NaN         NaN       NaN  ...        Aaron   
2        Mobile     1984.0         8.0      16.0  ...        Aaron   
3        Orange        NaN         NaN       NaN  ...         Aase   
4    Palm Beach        NaN         NaN       NaN  ...         Abad   
5     La Romana        NaN         NaN       NaN  ...         Abad   
6  Philadelphia     1905.0         5.0      17.0  ...       Abadie   
7       Latrobe     1957.0         1.0       6.0  ...  Abbaticchio   
8         Essex     1962.0         6.0      11.0  ...        Abbey   
9    Falls City     1926.0         4.0      27.0  ...        Abbey   

          nameGiven weight height bats throws       debut   finalGame  \
0       David Allan  215.0   75.0    R      R  2004-04-06  2015-08-23   
1       Henry Louis  180.0   72.0    R      R  1954-04-13  1976-10-03   
2        Tommie Lee  190.0   75.0    R      R  1962-04-10  1971-09-26   
3    Donald William  190.0   75.0    R      R  1977-07-26  1990-10-03   
4     Fausto Andres  184.0   73.0    L      L  2001-09-10  2006-04-13   
5  Fernando Antonio  220.0   73.0    L      L  2010-07-28  2019-09-28   
6           John W.  192.0   72.0    R      R  1875-04-26  1875-06-10   
7      Edward James  170.0   71.0    R      R  1897-09-04  1910-09-15   
8         Bert Wood  175.0   71.0    R      R  1892-06-14  1896-09-23   
9        Charles S.  169.0   68.0    L      L  1893-08-16  1897-08-19   

    retroID    bbrefID  
0  aardd001  aardsda01  
1  aaroh101  aaronha01  
2  aarot101  aaronto01  
3  aased001   aasedo01  
4  abada001   abadan01  
5  abadf001   abadfe01  
6  abadj101  abadijo01  
7  abbae101  abbated01  
8  abbeb101  abbeybe01  
9  abbec101  abbeych01  

[10 rows x 24 columns]

In [9]:
# Combine first name and last name
People["Name"] = People["nameFirst"].str.replace(" ", "") + " " + People["nameLast"]

The reason behind deleting spaces in the first name is to make sure that the initials are spaceless.    <br/>
For example, "C.J." and "C. J." will be different if we don't strip the spaces.

In [10]:
# Keep only Players born between 1970 and 1995
# Keep only Name, bbrefID to left join into FA_list
People_join = People[People.birthYear.isin(range(1970,1995))]
People_join = People_join[["Name", "bbrefID", "finalGame", "birthYear"]]
# Left join so that FA_list does not change
FA_list_joined = FA_list.merge(People_join, on = "Name", how = "left", indicator = True)

Set the range of players birth year so we limit the chances to get duplicate names and ID pairs.
Use finalGame and birthYear to further filter out players with same names. 

In [11]:
FA_list_joined.shape

(1147, 13)

In [12]:
FA_list.shape

(1118, 9)

There are 29 more rows in the joined dataframe than the original dataframe.

## Two problems:
1. Players without bbrefID (bbrefID = NaN) <br/> 
    - Some players have changed names while others have use different names to register.   <br/>
    - For example: B.J. Upton and Melvin Upton are the same player  <br/>
2. Players with multiple bbrefID (players with same names, hence some players have two bbrefID)  <br/>
    - This is becuase players have similar initials or the same names  <br/>
    - For example: There are two "Chirs Young". One is a pitcher while one in a outfielder. 

In [14]:
# Dataframe with players without bbrefID because the names are different in spotrac.com and baseball databank.
FA_list_joined[FA_list_joined["_merge"] == "left_only"]

Name Position  Age From   To  Years       Dollars  \
27      Jerry Hairston Jr.       LF   36  MIL  LAD      2    $6,000,000   
57         Vincent Padilla       RP   34  LAD  BOS      1    $1,500,000   
91             Mark Derosa       3B   37   SF  WSH      1      $800,000   
110           Melvin Upton       LF   28   TB  ATL      5   $72,500,000   
143         Mariano Rivera       RP   43  NYY  NYY      1   $10,000,000   
238            Mark Derosa       3B   38  WSH  TOR      1      $750,000   
249     Juan Carlos Oviedo       RP   31  MIA   TB      1      $490,000   
351     Juan Carlos Oviedo       RP   32   TB   TB      1    $1,500,000   
379       Yorbit Torrealba        C   35  COL  LAA      1      $950,000   
387         J.C. Gutiérrez       RP   30  LAA   SF      1      $850,000   
499      John Mayberry Jr.       RF   31  TOR  NYM      1    $1,450,000   
505         Eric Young Jr.       CF   30  NYM  ATL      1    $1,000,000   
509           Nick Massett       RP   33  COL  MIA      1    $1,000,000   
594         Norichika Aoki       LF   34   SF  SEA      1    $5,500,000   
599          Steven Pearce       1B   33  BAL   TB      1    $4,750,000   
655   Alberto Alburquerque       RP   30  DET  LAA      1    $1,100,000   
663             Matt Joyce       LF   31  LAA  PIT      1    $1,000,000   
668         Eric Young Jr.       CF   31  NYM  MIL      1    $1,000,000   
712           Michael Dunn       RP   32  MIA  COL      3   $19,000,000   
722          Steven Pearce       1B   34  BAL  TOR      2   $12,500,000   
732             Matt Joyce       LF   32  PIT  OAK      2   $11,000,000   
787             Tom Milone       SP   30  MIN  MIL      1    $1,250,000   
829          C.C. Sabathia       SP   37  NYY  NYY      1   $10,000,000   
846          Michael Fiers       SP   33  HOU  DET      1    $6,000,000   
870          Seung-Hwan Oh       RP   35  STL  TOR      1    $2,000,000   
914          D.J. LeMahieu       2B   30  COL  NYY      2   $24,000,000   
929          Michael Fiers       SP   34  OAK  OAK      2   $14,100,000   
1001          Zack Wheeler       SP   30  NYM  PHI      5  $118,000,000   
1093     Alexander Claudio       RP   28  MIL  MIL      1    $1,750,000   
1104            Matt Joyce       LF   35  ATL  MIA      1    $1,500,000   
1130          C.J. Edwards       RP   28   SD  SEA      1      $950,000   
1134           J.T. Riddle       SS   28  MIA  PIT      1      $850,000   
1138        Daniel Winkler       RP   30   SF  CHC      1      $750,000   

     Average_Salary  FA_year bbrefID finalGame  birthYear     _merge  
27       $3,000,000     2012     NaN       NaN        NaN  left_only  
57       $1,500,000     2012     NaN       NaN        NaN  left_only  
91         $800,000     2012     NaN       NaN        NaN  left_only  
110     $14,500,000     2013     NaN       NaN        NaN  left_only  
143     $10,000,000     2013     NaN       NaN        NaN  left_only  
238        $750,000     2013     NaN       NaN        NaN  left_only  
249        $490,000     2013     NaN       NaN        NaN  left_only  
351      $1,500,000     2014     NaN       NaN        NaN  left_only  
379        $950,000     2014     NaN       NaN        NaN  left_only  
387        $850,000     2014     NaN       NaN        NaN  left_only  
499      $1,450,000     2015     NaN       NaN        NaN  left_only  
505      $1,000,000     2015     NaN       NaN        NaN  left_only  
509      $1,000,000     2015     NaN       NaN        NaN  left_only  
594      $5,500,000     2016     NaN       NaN        NaN  left_only  
599      $4,750,000     2016     NaN       NaN        NaN  left_only  
655      $1,100,000     2016     NaN       NaN        NaN  left_only  
663      $1,000,000     2016     NaN       NaN        NaN  left_only  
668      $1,000,000     2016     NaN       NaN        NaN  left_only  
712      $6,333,333     2017     NaN       NaN        NaN  left_only  
722      $6,250,000     2017     NaN       NaN     

In [15]:
# Dataframe of players with multiple bbrefID
FA_list_joined[FA_list_joined[["Name", "FA_year"]].duplicated(keep=False)]

Name Position  Age From   To  Years       Dollars  \
2              Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
3              Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
32          Alex Gonzalez       SS   35  ATL  MIL      1    $4,250,000   
33          Alex Gonzalez       SS   35  ATL  MIL      1    $4,250,000   
34          Freddy Garcia       SP   35  NYY  NYY      1    $4,000,000   
35          Freddy Garcia       SP   35  NYY  NYY      1    $4,000,000   
80            Chris Young       RP   33  NYM  NYM      1      $900,000   
81            Chris Young       RP   33  NYM  NYM      1      $900,000   
159           Scott Baker       SP   31  MIN  CHC      1    $5,500,000   
160           Scott Baker       SP   31  MIN  CHC      1    $5,500,000   
187           Chris Young       RP   34  NYM  WSH      1    $2,000,000   
188           Chris Young       RP   34  NYM  WSH      1    $2,000,000   
204         Alex Gonzalez       SS   36  MIL  MIL      1    $1,450,000   
205         Alex Gonzalez       SS   36  MIL  MIL      1    $1,450,000   
294           Nelson Cruz       DH   33  TEX  BAL      1    $8,000,000   
295           Nelson Cruz       DH   33  TEX  BAL      1    $8,000,000   
300           Chris Young       RF   30  OAK  NYM      1    $7,250,000   
301           Chris Young       RF   30  OAK  NYM      1    $7,250,000   
323   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
324   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
355         Freddy Garcia       SP   37  ATL  ATL      1    $1,250,000   
356         Freddy Garcia       SP   37  ATL  ATL      1    $1,250,000   
372           Scott Baker       SP   32  CHC  SEA      1    $1,000,000   
373           Scott Baker       SP   32  CHC  SEA      1    $1,000,000   
413           Nelson Cruz       DH   34  BAL  SEA      4   $57,000,000   
414           Nelson Cruz       DH   34  BAL  SEA      4   $57,000,000   
434   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
435   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
492           Scott Baker       SP   33  TEX  NYY      1    $1,500,000   
493           Scott Baker       SP   33  TEX  NYY      1    $1,500,000   
532           Chris Young       RP   36  SEA   KC      1      $675,000   
533           Chris Young       RP   36  SEA   KC      1      $675,000   
569           Chris Young       RF   32  NYY  BOS      2   $13,000,000   
570           Chris Young       RF   32  NYY  BOS      2   $13,000,000   
574           Chris Young       RP   37   KC   KC      2   $11,500,000   
575           Chris Young       RP   37   KC   KC      2   $11,500,000   
619          Chris Carter       1B   29  HOU  MIL      1    $2,500,000   
620          Chris Carter       1B   29  HOU  MIL      1    $2,500,000   
689            Jose Reyes       SS   33  COL  NYM      1      $507,500   
690            Jose Reyes       SS   33  COL  NYM      1      $507,500   
754          Chris Carter       1B   30  MIL  NYY      1    $3,500,000   
755          Chris Carter       1B   30  MIL  NYY      1    $3,500,000   
850       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
851       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
852       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
871           Chris Young       RF   34  BOS  LAA      1    $2,000,000   
872           Chris Young       RF   34  BOS  LAA      1    $2,000,000   
874            Jose Reyes       SS   35  NYM  NYM      1    $2,000,000   
875            Jose Reyes       SS   35  NYM  NYM      1    $2,000,000   
927           Nelson Cruz       DH   38  SEA  MIN      1   $14,300,000   
928           Nelson Cruz       DH   38  SEA  MIN      1   $14,300,000   
1111          Luis Garcia       RP   33  LAA  TEX      1    $1,100,000   
1112          Luis Garcia       RP   33  LAA  TEX      1    $1,100,000   
1113          Luis Garcia

##  Players without bbrefID

There is no smart way but to fill in the IDs manually. 

In [16]:
# The index of NaN bbrefID in FA_list_joined
bbrefID_NA_index = FA_list_joined[FA_list_joined["bbrefID"].isnull()].index

In [17]:
# The names of NaN bbrefID in FA_list_joined
Name_NaNID = FA_list_joined[FA_list_joined["bbrefID"].isnull()].Name

In [18]:
# Fill in the NaN manually. 
NaNID = ["hairsje02", "padilvi01", "derosma01", "uptonbj01", "riverma01", "derosma01", "nunezle01", "nunezle01", "torreyo01", "gutieju01", "maybejo02", "younger03", "masseni01", "aokino01", "pearcst01", "albural01", "joycema01", "younger03", "dunnmi01","pearcst01","joycema01", "milonto01", "sabatc.01", "fiersmi01", "ohse01", "lemahdj01","fiersmi01", "wheelza01", "claudal01", "joycema01", "edwarca01", "riddljt01", "winklda01"]

In [19]:
len(NaNID) == len(Name_NaNID)

True

Through the process, I accidentally figured out row 523 Jeremy McBryde have not made it to the Majors yet, so will leave this row out of the analysis for now. There might be other players that have not played in the Majors but appears in the dataset. 

## Players with multiple bbrefID

In [20]:
# The dataframe of players with multiple bbrefID
FA_list_duplicated = FA_list_joined[FA_list_joined.drop(["bbrefID", "finalGame","birthYear","_merge"], axis = 1).duplicated(keep = False)]
FA_list_duplicated

Name Position  Age From   To  Years       Dollars  \
2              Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
3              Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
32          Alex Gonzalez       SS   35  ATL  MIL      1    $4,250,000   
33          Alex Gonzalez       SS   35  ATL  MIL      1    $4,250,000   
34          Freddy Garcia       SP   35  NYY  NYY      1    $4,000,000   
35          Freddy Garcia       SP   35  NYY  NYY      1    $4,000,000   
80            Chris Young       RP   33  NYM  NYM      1      $900,000   
81            Chris Young       RP   33  NYM  NYM      1      $900,000   
159           Scott Baker       SP   31  MIN  CHC      1    $5,500,000   
160           Scott Baker       SP   31  MIN  CHC      1    $5,500,000   
187           Chris Young       RP   34  NYM  WSH      1    $2,000,000   
188           Chris Young       RP   34  NYM  WSH      1    $2,000,000   
204         Alex Gonzalez       SS   36  MIL  MIL      1    $1,450,000   
205         Alex Gonzalez       SS   36  MIL  MIL      1    $1,450,000   
294           Nelson Cruz       DH   33  TEX  BAL      1    $8,000,000   
295           Nelson Cruz       DH   33  TEX  BAL      1    $8,000,000   
300           Chris Young       RF   30  OAK  NYM      1    $7,250,000   
301           Chris Young       RF   30  OAK  NYM      1    $7,250,000   
323   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
324   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
355         Freddy Garcia       SP   37  ATL  ATL      1    $1,250,000   
356         Freddy Garcia       SP   37  ATL  ATL      1    $1,250,000   
372           Scott Baker       SP   32  CHC  SEA      1    $1,000,000   
373           Scott Baker       SP   32  CHC  SEA      1    $1,000,000   
413           Nelson Cruz       DH   34  BAL  SEA      4   $57,000,000   
414           Nelson Cruz       DH   34  BAL  SEA      4   $57,000,000   
434   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
435   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
492           Scott Baker       SP   33  TEX  NYY      1    $1,500,000   
493           Scott Baker       SP   33  TEX  NYY      1    $1,500,000   
532           Chris Young       RP   36  SEA   KC      1      $675,000   
533           Chris Young       RP   36  SEA   KC      1      $675,000   
569           Chris Young       RF   32  NYY  BOS      2   $13,000,000   
570           Chris Young       RF   32  NYY  BOS      2   $13,000,000   
574           Chris Young       RP   37   KC   KC      2   $11,500,000   
575           Chris Young       RP   37   KC   KC      2   $11,500,000   
619          Chris Carter       1B   29  HOU  MIL      1    $2,500,000   
620          Chris Carter       1B   29  HOU  MIL      1    $2,500,000   
689            Jose Reyes       SS   33  COL  NYM      1      $507,500   
690            Jose Reyes       SS   33  COL  NYM      1      $507,500   
754          Chris Carter       1B   30  MIL  NYY      1    $3,500,000   
755          Chris Carter       1B   30  MIL  NYY      1    $3,500,000   
850       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
851       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
852       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
871           Chris Young       RF   34  BOS  LAA      1    $2,000,000   
872           Chris Young       RF   34  BOS  LAA      1    $2,000,000   
874            Jose Reyes       SS   35  NYM  NYM      1    $2,000,000   
875            Jose Reyes       SS   35  NYM  NYM      1    $2,000,000   
927           Nelson Cruz       DH   38  SEA  MIN      1   $14,300,000   
928           Nelson Cruz       DH   38  SEA  MIN      1   $14,300,000   
1111          Luis Garcia       RP   33  LAA  TEX      1    $1,100,000   
1112          Luis Garcia       RP   33  LAA  TEX      1    $1,100,000   
1113          Luis Garcia

There are 2 ways to help determine which bbrefID to keep (which row to keep):   <br/> 
1) Check finalGame: If the player played his final game before the FA contract, it would not make sense.   <br/>
2) Check age using birthYear: The age of signing should be the same as the year of signing minus birthYear.

In [21]:
# Make finalGame a year integer 
finalGame = list()
for i in FA_list_duplicated["finalGame"].index:
    finalGame.append(FA_list_duplicated["finalGame"][i][0:4])
finalGame_int = list()
for i in range(len(finalGame)):
    finalGame_int.append(int(finalGame[i]))
FA_list_duplicated.iloc[:,10] = finalGame_int

C:\Users\iamia\Anaconda3\lib\site-packages\pandas\core\indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [22]:
# Make birthYear and finalGame integer
birthYear_int = FA_list_duplicated["birthYear"].astype(int)
FA_list_duplicated.iloc[:,11] = birthYear_int
# Have to make an object and put it back

In [23]:
keep_finalGame = list()
age_difference = list()
for i in range(len(FA_list_duplicated)):
    keep_finalGame.append(FA_list_duplicated.iloc[i,10] > 2010)  # True if finalGame is later than 2010
    age_difference.append(FA_list_duplicated.iloc[i,8] - FA_list_duplicated.iloc[i,11] - FA_list_duplicated.iloc[i,2])

In [24]:
# Allow +- 1 difference in age and FA_year - birthDate
keep_Age = list()
for i in range(len(age_difference)):
    if age_difference[i] in [-1, 0, 1]:
        keep_Age.append(True)
    else: 
        keep_Age.append(False)

In [25]:
FA_list_duplicated["Keep_Age"] = keep_Age
FA_list_duplicated["Keep_finalGame"] = keep_finalGame

C:\Users\iamia\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\iamia\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [25]:
FA_list_duplicated

Name Position  Age From   To  Years       Dollars  \
2              Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
3              Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
32          Alex Gonzalez       SS   35  ATL  MIL      1    $4,250,000   
33          Alex Gonzalez       SS   35  ATL  MIL      1    $4,250,000   
34          Freddy Garcia       SP   35  NYY  NYY      1    $4,000,000   
35          Freddy Garcia       SP   35  NYY  NYY      1    $4,000,000   
80            Chris Young       RP   33  NYM  NYM      1      $900,000   
81            Chris Young       RP   33  NYM  NYM      1      $900,000   
159           Scott Baker       SP   31  MIN  CHC      1    $5,500,000   
160           Scott Baker       SP   31  MIN  CHC      1    $5,500,000   
187           Chris Young       RP   34  NYM  WSH      1    $2,000,000   
188           Chris Young       RP   34  NYM  WSH      1    $2,000,000   
204         Alex Gonzalez       SS   36  MIL  MIL      1    $1,450,000   
205         Alex Gonzalez       SS   36  MIL  MIL      1    $1,450,000   
294           Nelson Cruz       DH   33  TEX  BAL      1    $8,000,000   
295           Nelson Cruz       DH   33  TEX  BAL      1    $8,000,000   
300           Chris Young       RF   30  OAK  NYM      1    $7,250,000   
301           Chris Young       RF   30  OAK  NYM      1    $7,250,000   
323   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
324   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
355         Freddy Garcia       SP   37  ATL  ATL      1    $1,250,000   
356         Freddy Garcia       SP   37  ATL  ATL      1    $1,250,000   
372           Scott Baker       SP   32  CHC  SEA      1    $1,000,000   
373           Scott Baker       SP   32  CHC  SEA      1    $1,000,000   
413           Nelson Cruz       DH   34  BAL  SEA      4   $57,000,000   
414           Nelson Cruz       DH   34  BAL  SEA      4   $57,000,000   
434   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
435   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
492           Scott Baker       SP   33  TEX  NYY      1    $1,500,000   
493           Scott Baker       SP   33  TEX  NYY      1    $1,500,000   
532           Chris Young       RP   36  SEA   KC      1      $675,000   
533           Chris Young       RP   36  SEA   KC      1      $675,000   
569           Chris Young       RF   32  NYY  BOS      2   $13,000,000   
570           Chris Young       RF   32  NYY  BOS      2   $13,000,000   
574           Chris Young       RP   37   KC   KC      2   $11,500,000   
575           Chris Young       RP   37   KC   KC      2   $11,500,000   
619          Chris Carter       1B   29  HOU  MIL      1    $2,500,000   
620          Chris Carter       1B   29  HOU  MIL      1    $2,500,000   
689            Jose Reyes       SS   33  COL  NYM      1      $507,500   
690            Jose Reyes       SS   33  COL  NYM      1      $507,500   
754          Chris Carter       1B   30  MIL  NYY      1    $3,500,000   
755          Chris Carter       1B   30  MIL  NYY      1    $3,500,000   
850       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
851       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
852       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
871           Chris Young       RF   34  BOS  LAA      1    $2,000,000   
872           Chris Young       RF   34  BOS  LAA      1    $2,000,000   
874            Jose Reyes       SS   35  NYM  NYM      1    $2,000,000   
875            Jose Reyes       SS   35  NYM  NYM      1    $2,000,000   
927           Nelson Cruz       DH   38  SEA  MIN      1   $14,300,000   
928           Nelson Cruz       DH   38  SEA  MIN      1   $14,300,000   
1111          Luis Garcia       RP   33  LAA  TEX      1    $1,100,000   
1112          Luis Garcia       RP   33  LAA  TEX      1    $1,100,000   
1113          Luis Garcia

In [26]:
print(FA_list_duplicated.shape)
print(FA_list_duplicated[FA_list_duplicated["Keep_Age"] & FA_list_duplicated["Keep_finalGame"]].shape)

(56, 15)
(30, 15)


In [27]:
# Keep those with two trues
FA_list_duplicated[FA_list_duplicated["Keep_Age"] & FA_list_duplicated["Keep_finalGame"]]

Name Position  Age From   To  Years       Dollars  \
2              Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
33          Alex Gonzalez       SS   35  ATL  MIL      1    $4,250,000   
35          Freddy Garcia       SP   35  NYY  NYY      1    $4,000,000   
80            Chris Young       RP   33  NYM  NYM      1      $900,000   
160           Scott Baker       SP   31  MIN  CHC      1    $5,500,000   
187           Chris Young       RP   34  NYM  WSH      1    $2,000,000   
205         Alex Gonzalez       SS   36  MIL  MIL      1    $1,450,000   
295           Nelson Cruz       DH   33  TEX  BAL      1    $8,000,000   
301           Chris Young       RF   30  OAK  NYM      1    $7,250,000   
323   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
324   Francisco Rodriguez       RP   32  BAL  MIL      1    $3,250,000   
356         Freddy Garcia       SP   37  ATL  ATL      1    $1,250,000   
373           Scott Baker       SP   32  CHC  SEA      1    $1,000,000   
414           Nelson Cruz       DH   34  BAL  SEA      4   $57,000,000   
434   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
435   Francisco Rodriguez       RP   33  MIL  MIL      2   $13,000,000   
493           Scott Baker       SP   33  TEX  NYY      1    $1,500,000   
532           Chris Young       RP   36  SEA   KC      1      $675,000   
570           Chris Young       RF   32  NYY  BOS      2   $13,000,000   
574           Chris Young       RP   37   KC   KC      2   $11,500,000   
620          Chris Carter       1B   29  HOU  MIL      1    $2,500,000   
689            Jose Reyes       SS   33  COL  NYM      1      $507,500   
755          Chris Carter       1B   30  MIL  NYY      1    $3,500,000   
850       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
852       Miguel Gonzalez       SP   34  TEX  CHW      1    $4,750,000   
872           Chris Young       RF   34  BOS  LAA      1    $2,000,000   
874            Jose Reyes       SS   35  NYM  NYM      1    $2,000,000   
928           Nelson Cruz       DH   38  SEA  MIN      1   $14,300,000   
1113          Luis Garcia       RP   33  LAA  TEX      1    $1,100,000   
1119           Matt Duffy       3B   29   TB  TEX      1    $1,000,000   

     Average_Salary  FA_year    bbrefID  finalGame  birthYear _merge  \
2       $17,666,667     2012  reyesjo01       2018       1983   both   
33       $4,250,000     2012  gonzaal02       2014       1977   both   
35       $4,000,000     2012  garcifr03       2013       1976   both   
80         $900,000     2012  youngch03       2017       1979   both   
160      $5,500,000     2013  bakersc02       2015       1981   both   
187      $2,000,000     2013  youngch03       2017       1979   both   
205      $1,450,000     2013  gonzaal02       2014       1977   both   
295      $8,000,000     2014   cruzne02       2019       1980   both   
301      $7,250,000     2014  youngch04       2018       1983   both   
323      $3,250,000     2014  rodrifr03       2017       1982   both   
324      $3,250,000     2014  rodrifr04       2011       1983   both   
356      $1,250,000     2014  garcifr03       2013       1976   both   
373      $1,000,000     2014  bakersc02       2015       1981   both   
414     $14,250,000     2015   cruzne02       2019       1980   both   
434      $6,500,000     2015  rodrifr03       2017       1982   both   
435      $6,500,000     2015  rodrifr04       2011       1983   both   
493      $1,500,000     2015  bakersc02       2015       1981   both   
532        $675,000     2015  youngch03       2017       1979   both   
570      $6,500,000     2016  youngch04       2018       1983   both   
574      $5,750,000     2016  youngch03       2017       1979   both   
620      $2,500,000     2016  cartech02       2017       1986   both   
689        $507,500     2016  reyesjo01       2018       1983   both   
755      $3,500,000     2017  cartech02       2017       1986   both  

There are only 3 pairs of duplicates that we have to handle manually: Francisco Rodriguez(x2) and Miguel Gonzalez.  <br/>
By manually checking, bbrefID for Francisco Rodriguez is rodrifr03 and bbrefID for Miguel Gonzalez is gonzami03.   <br/>
This means index 324, 435, and 852 are false. 

In [28]:
FA_list_duplicated.shape

(56, 15)

In [29]:
# The index of rows we should keep
Keep_Index = FA_list_duplicated[FA_list_duplicated["Keep_Age"] & FA_list_duplicated["Keep_finalGame"]].index

In [30]:
len(Keep_Index)

30

In [31]:
# Drop the 3 manually.
Keep_Index = Keep_Index.drop([324, 435, 852])

In [32]:
NA_Index = list()
for i in FA_list_duplicated.index:
    NA_Index.append(i not in Keep_Index)
# Returns True if it should be dropped.
# A list of True/False

In [33]:
FA_list_duplicated.index[NA_Index] # The index that should be dropped from FA_list

Int64Index([   3,   32,   34,   81,  159,  188,  204,  294,  300,  324,  355,
             372,  413,  435,  492,  533,  569,  575,  619,  690,  754,  851,
             852,  871,  875,  927, 1111, 1112, 1120],
           dtype='int64')

In [34]:
len(FA_list_duplicated.index[NA_Index])

29

## Put back together 

In [35]:
len(bbrefID_NA_index)

33

In [36]:
# Fill in the NaNs in bbrefID
for i in range(len(NaNID)):
    FA_list_joined.at[bbrefID_NA_index[i], 'bbrefID'] = NaNID[i]

In [37]:
FA_list_joined = FA_list_joined.drop(FA_list_duplicated.index[NA_Index])
# Now FA_list_joined has the same number of rows as FA_list

In [38]:
FA_list_joined.shape

(1118, 13)

In [39]:
# Remove last three column 
FA_list_joined.drop(["finalGame", "birthYear", "_merge"], axis = 1, inplace = True)
FA_list_joined

Name Position  Age From   To  Years       Dollars  \
0      Albert Pujols       DH   32  STL  LAA     10  $240,000,000   
1     Prince Fielder       DH   28  MIL  DET      9  $214,000,000   
2         Jose Reyes       SS   29  NYM  MIA      6  $106,000,000   
4        C.J. Wilson       SP   31  TEX  LAA      5   $77,500,000   
5       Mark Buehrle       SP   33  CHW  MIA      4   $58,000,000   
...              ...      ...  ...  ...  ...    ...           ...   
1142  Dustin Garneau        C   32  OAK  HOU      1      $650,000   
1143   Collin McHugh       SP   33  HOU  BOS      1      $600,000   
1144      Jesse Hahn       RP   30   KC   KC      1      $600,000   
1145      Luke Maile        C   29  TOR  PIT      1      $583,500   
1146    Wei-Yin Chen       RP   35  MIA  SEA      1      $563,500   

     Average_Salary  FA_year    bbrefID  
0       $24,000,000     2012  pujolal01  
1       $23,777,778     2012  fieldpr01  
2       $17,666,667     2012  reyesjo01  
4       $15,500,000     2012  wilsocj01  
5       $14,500,000     2012  buehrma01  
...             ...      ...        ...  
1142       $650,000     2020  garnedu01  
1143       $600,000     2020  mchugco01  
1144       $600,000     2020   hahnje01  
1145       $583,500     2020  mailelu01  
1146       $563,500     2020   chenwe02  

[1118 rows x 10 columns]

In [40]:
# Take away the dollar sign and comma in Dollars and Average_Salary
FA_list_joined["Dollars"] = FA_list_joined["Dollars"].str.replace("$", "")
FA_list_joined["Dollars"] = FA_list_joined["Dollars"].str.replace(",", "")
FA_list_joined["Average_Salary"] = FA_list_joined["Average_Salary"].str.replace("$", "")
FA_list_joined["Average_Salary"] = FA_list_joined["Average_Salary"].str.replace(",", "")

In [26]:
FA_list_joined.head()

Name Position  Age From   To  Years       Dollars Average_Salary  \
0   Albert Pujols       DH   32  STL  LAA     10  $240,000,000    $24,000,000   
1  Prince Fielder       DH   28  MIL  DET      9  $214,000,000    $23,777,778   
2      Jose Reyes       SS   29  NYM  MIA      6  $106,000,000    $17,666,667   
3      Jose Reyes       SS   29  NYM  MIA      6  $106,000,000    $17,666,667   
4     C.J. Wilson       SP   31  TEX  LAA      5   $77,500,000    $15,500,000   

   FA_year    bbrefID   finalGame  birthYear _merge  
0     2012  pujolal01  2019-09-29     1980.0   both  
1     2012  fieldpr01  2016-07-18     1984.0   both  
2     2012  reyesjo01  2018-09-30     1983.0   both  
3     2012  reyesjo02  2006-09-30     1983.0   both  
4     2012  wilsocj01  2015-07-28     1980.0   both

2 changes are required during the analysis 
1) Tyler Colvin (colvity01) should be an outfielder instead of a RP
2) Get rid of Tony Barnettes 2016 FA becuase he has not pitched in the Majors before 2016

In [46]:
FA_list_joined.iloc[353,1] = "OF"
FA_list_joined = FA_list_joined.drop(611)

In [47]:
# Output data
FA_list_joined.to_csv("FA_list_with_bbrefID.csv", index = False)